algorithms/finetune/cal_ql.py

```diff
@@ -0,0 +1,1143 @@
+        "actor": actor,
+        "actor_optimizer": actor_optimizer,
+        "discount": config.discount,
+        "soft_target_update_rate": config.soft_target_update_rate,
+        "device": config.device,
+        # CQL
+        "target_entropy": -np.prod(env.action_space.shape).item(),
+        "alpha_multiplier": config.alpha_multiplier,
+        "use_automatic_entropy_tuning": config.use_automatic_entropy_tuning,
+        "backup_entropy": config.backup_entropy,
+        "policy_lr": config.policy_lr,
+        "qf_lr": config.qf_lr,
+        "bc_steps": config.bc_steps,
+        "target_update_period": config.target_update_period,
+        "cql_n_actions": config.cql_n_actions,
+        "cql_importance_sample": config.cql_importance_sample,
+        "cql_lagrange": config.cql_lagrange,
+        "cql_target_action_gap": config.cql_target_action_gap,
+        "cql_temp": config.cql_temp,
+        "cql_alpha": config.cql_alpha,
+        "cql_max_target_backup": config.cql_max_target_backup,
+        "cql_clip_diff_min": config.cql_clip_diff_min,
+        "cql_clip_diff_max": config.cql_clip_diff_max,
+    }
+
+    print("---------------------------------------")
+    print(f"Training Cal-QL, Env: {config.env}, Seed: {seed}")
+    print("---------------------------------------")
+
+    # Initialize actor
+    trainer = CalQL(**kwargs)
+
+    if config.load_model != "":
+        policy_file = Path(config.load_model)
+        trainer.load_state_dict(torch.load(policy_file))
+        actor = trainer.actor
+
+    wandb_init(asdict(config))
+
+    evaluations = []
+    state, done = env.reset(), False
+    episode_return = 0
+    episode_step = 0
+
+    print("Offline pretraining")
+    for t in range(int(config.offline_iterations) + int(config.online_iterations)):
+        if t == config.offline_iterations:
+            print("Online tuning")
+            trainer.switch_calibration()
+            trainer.cql_alpha = config.cql_alpha_online
+        online_log = {}
+        if t >= config.offline_iterations:
+            episode_step += 1
+            action, _ = actor(
+                torch.tensor(
+                    state.reshape(1, -1), device=config.device, dtype=torch.float32
+                )
+            )
+            action = action.cpu().data.numpy().flatten()
+            next_state, reward, done, _ = env.step(action)
+
+            episode_return += reward
+            real_done = False  # Episode can timeout which is different from done
+            if done and episode_step < max_steps:
+                real_done = True
+
+            if config.normalize_reward:
+                reward = modify_reward_online(reward, config.env, **reward_mod_dict)
+            online_buffer.add_transition(state, action, reward, next_state, real_done)
+            state = next_state
+
+            if done:
+                state, done = env.reset(), False
+
+                online_log["episode_return"] = episode_return
```

Comment:
add prefix train/

use equivalent names for the same metrics for both evaluation and training curves